File: day-1/go/solution_2.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	data, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}

	lines := strings.Split(string(data), "\r\n")
	mostCalories := make([]int, 3)

	calories := make([]int, 0)

	for _, line := range lines {
		if line == "" {
			sum := 0
			for _, calorie := range calories {
				sum += calorie
			}
			leastCalorieIndex := 0
			for index, mostCalorie := range mostCalories {
				if mostCalorie < mostCalories[leastCalorieIndex] {
					leastCalorieIndex = index
				}
			}
			if sum > mostCalories[leastCalorieIndex] {
				mostCalories[leastCalorieIndex] = sum
			}
			calories = make([]int, 0)
			continue
		}
		number, err := strconv.Atoi(line)
		if err != nil {
			panic(err)
		}
		calories = append(calories, number)
	}

	fmt.Printf("The elves carrying the most calories carry %d, %d and %d calories\n", mostCalories[0], mostCalories[1], mostCalories[2])
	fmt.Printf("Their combined calories is: %d\n", mostCalories[0]+mostCalories[1]+mostCalories[2])
}

File: day-1/rust/src/solution_2.rs
use std::fs;

fn main() {
    let file_content = fs::read_to_string("input.txt").unwrap();
    let lines = file_content.lines();
    let mut max_calories = [0, 0, 0];
    let mut sum = 0;
    for line in lines {
        if line.is_empty() {
            let mut least_calorie_index = 0;
            for (index, max_calorie) in max_calories.iter().enumerate() {
                if max_calorie < &max_calories[least_calorie_index] {
                    least_calorie_index = index;
                }
            }
            if sum > max_calories[least_calorie_index] {
                max_calories[least_calorie_index] = sum;
            }
            sum = 0;
            continue;
        }
        let number: i32 = line.parse().unwrap();
        sum += number;
    }
    println!(
        "The elves carrying the most calories carry {}, {}, and {} calories",
        max_calories[0], max_calories[1], max_calories[2]
    );
    println!(
        "Their cumulative calories is {}",
        max_calories.iter().map(|c| c).sum::<i32>()
    );
}
